README.md
# AutoPlay PokeMeow Discord 🤖
The timings between each catchs and fishs are defined line 59 for the fishes and line 65 ( both in `bot.py`) for classic pokemons. The bare minimum is 10seconds (or Pokemeow will tell you to wait).
I would advise you to set it to a random number by replacing the `10` by something like `randint(10, 20)` (10 will be the minimum waiting time and 20 the maximum).

## Supported features
This bot can currently handle :
1. All the checklist commands (so `;daily`, `;hunt` etc...)
// NO VOTE ANYMORE BECAUSE CAPTCHA
2. Vote for Pokemeow every 12H (can fail sometimes, top.gg added a captcha)
// NO VOTE ANYMORE BECAUSE CAPTCHA
3. Throw Pokeballs depending on the rarity of the Pokemon
4. Buy Pokeballs when you're out of stock (only if you can buy a certain amount of each less effective balls : see dedicated section)
5. Handle your eggs
6. Use all your Repels and Golden Razz Berries
7. Open every boxes in your bag
8. Can fish for you (via an argument)

## How buying works 
With the default settings, the bot will buy :
- 1 Masterball only if you have more than 109K coins
- 3 Ultraballs if > 9K
- 5 Superballs if > 4500
- 5 Pokeballs if > 1k

⚠️ So if it didn't bought any balls it means that you were a little bit too poor ⚠️ 

This is a security to avoid situations where you buy 1 Masterball at 100K and don't have enought coins left to buy lesser effectives balls.

## Installating

### On your pc
1. Install your favorite version of python 3 and add it to PATH.
2. If not done already download this repo on your computer as zip and unzip it.
3. Inside of the repo type the following command `python -m pip install -r requirments.txt` (If your on a linux distribution or MacOS you got a specify the python version so it's `python3 ...`)

### On your discord
4. a. You got to activate your Pokemeow account (type `;p` at least once)

   b. I recommand you to have at least 1,000 Coins.


## Launching 

In order to launch the bot you have to type the following command: `python bot.py` (`python3 bot.py` if you're on Linux or MACos) with some required arguments :
- `-U` : Your displayed username on the discord server you wan't to use this bot (put it inside quotes it it contains a space or if your not sure).
- `-M` : Your Discord mail adress.
- `-P` : Your Discord Password.
- `-C` : The direct link to the channel of your choice (between quotes just in case :wink:)

The following arguments depends on your personnal setup, and are optionnal : 
- `-F` : If you want this program to fish for you.
- `-FA` : If you got the 2FA authetification, you have to put the 2FA key. Because 2FA is designed to prevent from bots this program won't be able to run in headless mode and you'll have to type manually the code that the bot will give you.
- `-H` : For Headless mode (without the Browser window), ⚠️ This will not work if you have activate 2FA !
- `-D` : If you want to use another driver, Firefox is the default one. Add 'Chrome' after the flag in order to use Chrome.
- `-BACKWARDS` : If the bot is typing backwards, this will reverse everything the bot is typing on discord.

So to load the script you have to type in your Terminal or CMD some thing like : 
`python bot.py -U 'John Doe' -M 'johndoe@gmail.com' -P 'mySuperPassword1234' -C 'CHANNEL LINK'` and every other flag needed for your use case.
(or `python3` for Linux or MacOS)

⚠️ You can of course create a .sh or .bat file in order to execute the command without repetively type the command but rememeber this is a big security breach and we're not responsive for any dataleaking ⚠️

## Stoping
To stop the program simply type Ctrl + C in the command promt, or close the browser if not in Headless mode, or close directly the command promt.


## Adding new webdrivers
If you want youre favorite browser to be supported, visit : https://www.selenium.dev/documentation/en/getting_started_with_webdriver/browsers/ and see if your's part of the supported drivers of Selenium.


## Disclamer ⚠️
- We're of course not responsible for any ban you recieve for using this bot.
- For now only Firefox, Opera, Edge and Chrome are supported.
- Edge is only supported on Windows.
- Running multiple bots in one channel is definitively not recommanded.


CHANGELOG.md
- `23 May 2021` : Workaround Backwards typing by adding `-BACKWARDS` flag (issue #10).
- `21 Mar 2021` : Fix pokeballs not being bought after fishes and before pokemon (issue #7).
- `11 Mar 2021` : Remove Vote support because can't bypass capthca on top.gg.
- `26 Feb 2021` : Added explanation on how buying works in the README file.

bot.py
from random import randint
from sys import argv
from traceback import format_exc
from argparse import ArgumentParser

from Classes.Fishing import Fish
from Classes.Bag import Bag
from Classes.Pokemon import Pokemon
from Classes.Checklists import Checklists
from Classes.Feedback import Feedback
from Classes.CustomDriver import CustomDriver

"""
TODO
  - captcha ?
"""
ARGS = {
  '-C': {'metavar': 'Channel link', 'help': 'Put the CHANNEL link (between quotes) you want to connect to.'},
  '-U': {'metavar': 'USERNAME', 'help': 'Put your discord USERNAME, between quotes if it has spaces in it.'},
  '-M': {'metavar': 'MAIL', 'help': 'Put your MAIL for connecting discord.'},
  '-P': {'metavar': 'PASSWORD', 'help': 'Put your PASSWORD for connecting discord.'}
}
parser = ArgumentParser(description='A Bot to play Pokemeow on Discord')

for arg in ARGS: parser.add_argument(arg, required=True, type=str, help=ARGS[arg]['help'], metavar=ARGS[arg]['metavar'])

parser.add_argument('-D', metavar='Browser Name', type=str, help='Put your browser name. (ex: "Chrome")', default='Firefox')
parser.add_argument('-FA', metavar='2FA key', type=str, help='Put your 2FA key, if you got one.')
parser.add_argument('-H', action='store_true', help='Set this flag if you want to run the BROWSER in with no GUI')
parser.add_argument('-F', action='store_true', help='Set this flag if you want to fish (will not disable classic pokemons')
parser.add_argument('-BACKWARDS', action='store_true', help='Set this flag if the bot is typing backwards...')

ARGUMENTS = parser.parse_args()

try:
    BAKSUCESS = None

    FEEDBACK  = Feedback(ARGUMENTS.F)
    DRIVER    = CustomDriver(ARGUMENTS)
    BAG       = Bag(DRIVER)
    CHECKLIST = Checklists(DRIVER)

    while True:

        if not FEEDBACK.get_total('success') % 20 and FEEDBACK.get_total('success') != BAKSUCESS:
            BAKSUCESS = FEEDBACK.get_total('success')
            BAG.get_bag()
            if BAG.open_boxes(): BAG.get_bag()
            BAG.do_consumables()
            BAG.do_eggs()
            print(BAG)

        BAG.reassort()
        CHECKLIST.verify_checklist()

        if ARGUMENTS.F :
            fish = Fish(BAG, DRIVER)
            FEEDBACK.add_fish(fish)
            print(fish)
            DRIVER.attente( 10 , 'next catch')
            BAG.reassort()

        pkmn = Pokemon(BAG, DRIVER)
        FEEDBACK.add_pkmn(pkmn)
        print(pkmn)

        DRIVER.attente( 10 , f'next { "fish" if ARGUMENTS.F else "catch" }.')
        print('##########################################')

except KeyboardInterrupt: pass
except RecursionError: print('MAX RECURSION REACHED')
except: print(format_exc()) # pass

try: 
    print(BAG)
    print(FEEDBACK)
except: pass

try: DRIVER.quit()
except: pass
input('Press Enter to close !!')


Classes/Bag.py
from re import sub

class Bag():

    BALLS = {
            "Masterballs"  : {"call":  "mb", "id_shop":    4, "stock": 0, "to_buy":    1, "price": 100000}, # buy only if money > 109 000 coins
            "Premier balls": {"call": "prb", "id_shop": None, "stock": 0, "to_buy": None, "price":   None},
            "Ultraballs"   : {"call":  "ub", "id_shop":    3, "stock": 0, "to_buy":    3, "price":   1500}, # buy only if money >   9 000 coins
            "Greatballs"   : {"call":  "gb", "id_shop":    2, "stock": 0, "to_buy":    5, "price":    500}, # buy only if money >   4 500 coins
            "Pokeballs"    : {"call":  "pb", "id_shop":    1, "stock": 0, "to_buy":    5, "price":    200}, # buy only if money >   1 000 coins
            # "Diveballs"   : {"call":  "db", "id_shop": None, "stock": 0, "to_buy": None, "price":   None}
    }

    BOXES = {
        "Rare boxes":       {"call":  ";open rb", "stock": 0},
        "Super Rare boxes": {"call": ";open srb", "stock": 0},
        "Legendary boxes":  {"call":  ";open lb", "stock": 0},
        "Shiny boxes":      {"call":  ";open sb", "stock": 0},
    }

    LOOTBOXES = { "stock": 0, "expected": " opened", "call": ";lb all"}

    CONSUMABLES = { 
        'Golden Razz Berries' : {"stock": 0, "expected":   ", you", "call": ";grazz all"}, 
        'Repels'              : {"stock": 0, "expected":   ", you", "call": ";repels all"}
    }

    EGGS = {
        "stock": 0,
        "hatch": False,
        "hold" : False
    }

    def __init__(self, driver):
        self.DRIVER = driver

    def __repr__(self) -> str:
        balls = ( '\n'. join( f"""        ║{ f"{ self.BALLS[ball]['stock']:4d} : { ball:<12s}":^25s}║""" for ball in self.BALLS))
        return f"""
        ╔═════════════════════════╗
        ║   Your item inventory   ║
        ╠═════════════════════════╣
        ║        Currencies       ║
        ║{ f'{ self.money:^9,d} : PokeCoins':^25s}║ 
        ╟━━━━━━━━━━━━━━━━━━━━━━━━━╢
        ║          Balls          ║\n{ balls }
        ╟━━━━━━━━━━━━━━━━━━━━━━━━━╢
        ║       Catch Items       ║
        ║{ f"{ self.EGGS['stock'] } : Eggs":^25s}║
        ╚══━━━━═══════════════════╝
        """

    def open_boxes(self) -> bool:
        OPENED = False
        for box in self.BOXES:
            while self.BOXES[box]['stock']:
                OPENED = True
                self.DRIVER.WaitNew(self.BOXES[box]['call'], f"{ self.DRIVER.USERNAME } opened")
                self.BOXES[box]['stock'] -= 1
                self.DRIVER.attente(10, 'opening another box or continue..')
        
        if self.LOOTBOXES['stock']:
            self.DRIVER.WaitNew(self.LOOTBOXES['call'], f'{ self.DRIVER.USERNAME }{ self.LOOTBOXES["expected"] }')
            self.LOOTBOXES['stock'] = 0
            OPENED = True
            self.DRIVER.attente(5, 'ending boxes opening..')
        return OPENED

    def reassort(self) -> None:
        for index, ball in enumerate(self.BALLS.keys()):
            if self.BALLS[ball]['to_buy']    and \
               self.BALLS[ball]['stock'] < 1 and \
               self.money >= sum( self.BALLS[ball]['price'] * self.BALLS[ball]['to_buy'] if self.BALLS[ball]['to_buy'] else 0 for ball in list(self.BALLS.keys())[index:]) + (self.BALLS["Pokeballs"]['price'] * self.BALLS["Pokeballs"]['to_buy'] if ball !='Pokeballs' else 0): 
                print(f"Buying {self.BALLS[ball]['to_buy']} {ball} !")
                self.DRIVER.WaitNew(f';shop buy { self.BALLS[ball]["id_shop"] } { self.BALLS[ball]["to_buy"] }', f'{ self.DRIVER.USERNAME }, you bought')

                self.BALLS[ball]["stock"] += self.BALLS[ball]["to_buy"]
                self.money -= (self.BALLS[ball]['to_buy'] * self.BALLS[ball]['price'])
                self.DRIVER.attente(10, 'next catch or buy..')

    def update_balls(self, msg: str, CAPTURED: bool) -> None:
        if CAPTURED:
             self.money += int(sub(r'(.|\n)*?(\d?[0-9,]+) PokeCoins(.|\n)*' , r'\2', msg).replace(',', ''))

        for ball in self.BALLS:
            if ball in msg:
                self.BALLS[ball]['stock'] = int(sub(rf'(.|\n)*{ ball } ?: (\d?[0-9,]+)(.|\n)*', r'\2', msg).replace(',', ''))

    def get_bag(self) -> None:
        message = self.DRIVER.WaitNew(';items', f"{ self.DRIVER.USERNAME }'s item inventory")

        self.money = int(sub(r'(.|\n)*?(\d?[0-9,]+)x PokeCoins(.|\n)*' , r'\2', message.text).replace(',', ''))
        for ball in self.BALLS:
            self.BALLS[ball]['stock'] = int(sub(rf'(.|\n)*?(\d?[0-9,]+)x { ball }(.|\n)*'  , r'\2', message.text).replace(',', ''))

        for box in self.BOXES:
            self.BOXES[box]['stock'] = int(sub(rf'(.|\n)*?(\d?[0-9,]+)x { box }(.|\n)*'  , r'\2', message.text).replace(',', ''))

        for consumable in self.CONSUMABLES:
            self.CONSUMABLES[consumable]['stock'] = int(sub(rf'(.|\n)*?(\d?[0-9,]+)x { consumable }(.|\n)*'  , r'\2', message.text).replace(',', ''))

        self.LOOTBOXES['stock'] = int(sub(rf'(.|\n)*?(\d?[0-9,]+)x Lootboxes(.|\n)*'  , r'\2', message.text).replace(',', ''))

        self.EGGS['stock'] = int(sub(rf'(.|\n)*?(\d?[0-9,]+)x Eggs(.|\n)*'  , r'\2', message.text).replace(',', ''))
        self.EGGS['hold']  = 'COUNTER' in message.text
        self.EGGS['hatch'] = 'READY' in message.text

    def do_consumables(self) -> None:
        for consumable in self.CONSUMABLES:
            if self.CONSUMABLES[consumable]['stock']:
                self.DRIVER.WaitNew(self.CONSUMABLES[consumable]['call'], f'{ self.DRIVER.USERNAME }{ self.CONSUMABLES[consumable]["expected"] }')
                self.CONSUMABLES[consumable]['stock'] = 0
                self.DRIVER.attente(10, 'using another consumable or continue..')

    def do_eggs(self) -> None:
        if not self.EGGS['hold'] and self.EGGS['stock']:
            if self.EGGS['hatch']:
                self.DRIVER.WaitNew(';egg hatch', 'hatched')
                print(f'An egg was hatched !')
                self.DRIVER.attente(15, 'before holding a new egg')
            self.DRIVER.WaitNew(';egg hold', 'holding')
            print(f'Your now holding an egg !')
            self.EGGS['stock'] -= 1
            self.DRIVER.attente(10, 'next action..')
        
        elif self.EGGS['hatch']:
            self.DRIVER.WaitNew(';egg hatch', 'hatched')
            print(f'An egg was hatched !')
            self.DRIVER.attente(10, 'next action..')

Classes/Fishing.py
from time import time, sleep
from re import sub

class Fish():
    
    TIMEOUT = 15
    CAPTURED = None
    NAME = None
    BALL = 'Pokeballs' # default one in order to send greatballs

    LEGENDARYS = [
        # Gen I
        # Gen II
        "Suicune",
        # Gen III
        "Kyogre",
        # Gen IV
        "Palkia",
        "Phione",
        "Manaphy",
        # Gen V
        "Keldeo",
        "Volcanion",
        # Gen VI
        # Gen VII
        "Tapu Fini",
        # Gen VIII
        "Urshifu"
    ]

    def __init__(self, bag, driver):
        self.DRIVER = driver
        self.BAG = bag
        self.spawned, self.message, self.name = self.spawn(time())
        if self.spawned: 
            self.which_ball()
            self.CAPTURED = '\nCongratulations' in self.message.text
            self.BAG.update_balls(self.message.text, False) # False because not earning Pokecoins

    def spawn(self, START_TIME):
        message = self.DRIVER.WaitNew(';fish', f'{ self.DRIVER.USERNAME } cast')
        bak_txt = message.text
        while time() < START_TIME + self.TIMEOUT and bak_txt == message.text:
            sleep(0.1)
        if 'PULL' in message.text: # there is a fish
            self.DRIVER.WaitChangesOnMessage('pull', message)
            if 'got away' in message.text: return False, message, ''
            return True, message, sub( r'(.|\n)*?a wild (.*)!(.|\n)*', r'\2', message.text )
        elif 'got away' in message.text: return False, message, ''
        else: return None, message, ''


    def which_ball(self) -> None:
        if 'Shiny' in self.message.text or 'Golden' in self.message.text or any(contain in self.message.text for contain in self.LEGENDARYS): self.BALL = 'Masterballs'

        if self.BAG.BALLS[self.BALL]['stock'] > 0:
            return self.DRIVER.WaitChangesOnMessage(self.BAG.BALLS[self.BALL]['call'], self.message)
                    
        else:
            for i in range(list(self.BAG.BALLS.keys()).index(self.BALL), len(self.BAG.BALLS.keys())):
                if self.BAG.BALLS[list(self.BAG.BALLS.keys())[i]]['stock'] > 0:
                    return self.DRIVER.WaitChangesOnMessage(self.BAG.BALLS[list(self.BAG.BALLS.keys())[i]]['call'], self.message)
                            
            self.DRIVER.SendMessage('I\'m too poor !')

    def __repr__(self):
        return f"""It's a { self.name }, {'and is now in you PC !' if self.CAPTURED else 'ran away.'}""" if self.name else 'No Fish to see here.'

Classes/Pokemon.py
from re import sub

class Pokemon(): 

    RARITYS = {
        "Legendary"     :   "Masterballs",
        "Shiny"         :   "Masterballs",
        "Super Rare"    :    "Ultraballs",
        "Rare"          :    "Greatballs",
        "Uncommon"      :     "Pokeballs",
        "Common"        :     "Pokeballs"
    }

    CAPTURED = False

    def __init__(self, bag, driver) -> None:
        self.DRIVER = driver
        self.BAG = bag
        self.message, self.rarity, self.name = self.spawn()

        self.CAPTURED = self.which_ball()
        self.BAG.update_balls(self.message.text, self.CAPTURED)

    def __repr__(self) -> str:
        return f"""It's a {self.name}, it's rarity is {self.rarity}, {'and is now in you PC !' if self.CAPTURED else 'ran away.'}"""
 
    def spawn(self):
        message = self.DRIVER.WaitNew(';p', f"{ self.DRIVER.USERNAME } found")
        return ( message, 
                 sub( r'(.|\n)*?a wild .*!\n(.*) \((.|\n)*', r'\2', message.text ),  
                 sub( r'(.|\n)*?a wild (.*)!(.|\n)*', r'\2', message.text ) )

    def which_ball(self) -> None:
        for index, ball in enumerate(self.BAG.BALLS.keys()):
            if self.RARITYS.get(self.rarity) == ball: 
                if self.BAG.BALLS[ball]['stock'] > 0:
                    return self.DRIVER.WaitChangesOnMessage(self.BAG.BALLS[ball]['call'], self.message)
                    
                else:
                    for i in range(index, len(self.BAG.BALLS.keys())):
                        if self.BAG.BALLS[list(self.BAG.BALLS.keys())[i]]['stock'] > 0:
                            return self.DRIVER.WaitChangesOnMessage(self.BAG.BALLS[list(self.BAG.BALLS.keys())[i]]['call'], self.message)
                            
                    self.DRIVER.SendMessage('I\'m too poor !')

Classes/Feedback.py
from datetime import datetime

class Feedback():
    CAPTURED = {
        "Legendary"     : {'total': 0,'failed': 0, 'success': 0, 'SuccessNames': [], 'FailedNames': []},
        "Shiny"         : {'total': 0,'failed': 0, 'success': 0, 'SuccessNames': [], 'FailedNames': []},
        "Super Rare"    : {'total': 0,'failed': 0, 'success': 0},
        "Rare"          : {'total': 0,'failed': 0, 'success': 0},
        "Uncommon"      : {'total': 0,'failed': 0, 'success': 0},
        "Common"        : {'total': 0,'failed': 0, 'success': 0}
    }

    FISHS = {'total': 0,'failed': 0, 'success': 0, 'none': 0, 'SuccessNames': [], 'FailedNames': []}

    def __init__(self, FISH) -> None:
        self.FISH = FISH
        self.demarrage = datetime.now()
        print(f'It\'s { self.demarrage.strftime("%H:%M:%S") }, working time !')


    def add_pkmn(self, pkmn) -> None:
        for rarity in self.CAPTURED:
            if rarity == pkmn.rarity:
                if pkmn.CAPTURED:
                    self.CAPTURED[rarity]['success'] += 1
                    if rarity == 'Legendary' or rarity == 'Shiny': self.CAPTURED[rarity]['SuccessNames'].append(pkmn.name)
                else:
                    self.CAPTURED[rarity]['failed'] += 1
                    if rarity == 'Legendary' or rarity == 'Shiny': self.CAPTURED[rarity]['FailedNames'].append(pkmn.name)

                self.CAPTURED[rarity]['total'] += 1

    def add_fish(self, fish) -> None:
        if fish.CAPTURED:
            self.FISHS['success'] += 1
            if fish.BALL == 'Masterballs': self.FISHS['SuccessNames'].append(fish.name)
        elif fish.CAPTURED == None:
            self.FISHS['none'] += 1
        else:
            self.FISHS['failed'] += 1
            if fish.BALL == 'Masterballs': self.FISHS['FailedNames'].append(fish.name)
            
        self.FISHS['total'] += 1

    def get_total(self, cat: str) -> int: return sum(self.CAPTURED[rarity][cat] for rarity in self.CAPTURED) + self.FISHS['success']

    def get_duration(self):
        hours, remainder = divmod((datetime.now() - self.demarrage).seconds, 3600)
        minutes, seconds = divmod(remainder, 60)
        return f'{hours:02}H {minutes:02}m {seconds:02}s'

    def __repr__(self) -> str:
        caught = ( '\n'. join( f"""        ║{ f"{ self.CAPTURED[rarity]['success']:4d} : { rarity:<12s}":^22s}║""" for rarity in self.CAPTURED))
        missed = ( '\n'. join( f"""        ║{ f"{ self.CAPTURED[rarity]['failed']:4d} : { rarity:<12s}":^22s}║"""  for rarity in self.CAPTURED))
        total  = ( '\n'. join( f"""        ║{ f"{ self.CAPTURED[rarity]['total']:4d} : { rarity:<12s}":^22s}║"""   for rarity in self.CAPTURED))
        return f"""
        ╔══════════════════════╗
        ║       Duration       ║
        ║{self.get_duration():^22s}║ 
        ╟━━━━━━━━━━━━━━━━━━━━━━╢
        ║{ f"TOTAL FOUND : {self.get_total('total')}":^22s}║
        ╟----------------------╢\n{ total }
        ╟━━━━━━━━━━━━━━━━━━━━━━╢
        ╟       DETAILS        ╢
        ╟----------------------╢
        ║{ f"Caught : { self.get_total('success') }":^22s}║\n{ caught }
        ║----------------------║
        ║{ f"Missed : { self.get_total('failed') }":^22s}║\n{ missed }
        ╚══━━━━════════════════╝
        """ + (f"""
        ╔══════════════════════╗
        ║        Fishs         ║
        ║{ f"TOTAL TRIED : {self.FISHS['total']}":^22s}║
        ╟━━━━━━━━━━━━━━━━━━━━━━╢
        ╟       DETAILS        ╢
        ╟----------------------╢
        ║{ f"Didn't show up : { self.FISHS['none'] }":^22s}║
        ╟----------------------╢
        ║{ f"Caught : { self.FISHS['success'] }":^22s}║
        ║----------------------║
        ║{ f"Missed : { self.FISHS['failed'] }":^22s}║
        ╚══━━━━════════════════╝
        """ if self.FISH else '\r\nFishs weren\'t asked.\n')  + f"""
        \r{ f"Important fishes catched : { self.FISHS['SuccessNames'] }" if self.FISHS['SuccessNames'] else "" }
        \r{ f"Important fishes missed : { self.FISHS['FailedNames'] }" if self.FISHS['FailedNames'] else "" }
        \r{ f"Important Legendary catched : { self.CAPTURED['Legendary']['SuccessNames'] }" if self.CAPTURED['Legendary']['SuccessNames'] else "" }
        \r{ f"Important Legendary missed : { self.CAPTURED['Legendary']['FailedNames'] }" if self.CAPTURED['Legendary']['FailedNames'] else "" }
        \r{ f"Important Shiny catched : { self.CAPTURED['Shiny']['SuccessNames'] }" if self.CAPTURED['Shiny']['SuccessNames'] else "" } 
        \r{ f"Important Shiny missed : { self.CAPTURED['Shiny']['FailedNames'] }" if self.CAPTURED['Shiny']['FailedNames'] else "" } 
        """


Classes/CustomDriver.py
from decimal import Decimal
from sys import platform, stdout
from time import sleep

from pyotp import TOTP
from selenium import webdriver
from selenium.common.exceptions import TimeoutException
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options as ChromeOptions
from selenium.webdriver.edge.options import Options as EdgeOptions
from selenium.webdriver.firefox.options import Options as FirefoxOptions
from selenium.webdriver.opera.options import Options as OperaOptions
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait


class CustomDriver( webdriver.Firefox, 
                    webdriver.Opera,  
                    webdriver.Chrome,
                    webdriver.Edge):
    
    URL_VOTE_POKEMEOW = 'https://top.gg/bot/664508672713424926/vote'
    URL_LOGIN_TOP_GG = 'https://top.gg/login'
    AUTHORIZE_BUTTON_XPATH = "//button[@class='button-38aScr lookFilled-1Gx00P colorBrand-3pXr91 sizeMedium-1AC_Sl grow-q77ONN']"
    VOTE_BUTTON_XPATH = '//input[@class="voting btn btn-orange btn-4x vote-button"]'
    # <input id="votingvoted" type="button" class="voting btn btn-orange btn-4x vote-button" value="Vote">
    # MODIFIER = Keys.COMMAND if platform == "win32" or platform == 'linux' else Keys.COMMAND

    SEND_MSG_BAR_XPATH: str = "//div[@class='markup-2BOw-j slateTextArea-1Mkdgw fontSize16Padding-3Wk7zP']"
    SEND_MSG_BAR = None
    LAST_MSG = None
    NBR_MAX_RECURSION: int = 4

    MSG_XPATH: str = "//div[contains(@class,'message-2qnXI6 cozyMessage-3V1Y8y')]"
    POKEMEOW_USERNAME: str = 'PokéMeow'

    def __init__(self, ARGUMENTS):
        self.CHANNEL = ARGUMENTS.C
        self.USERNAME = ARGUMENTS.U
        self.MAIL = ARGUMENTS.M
        self.PASSWD = ARGUMENTS.P
        self.TOKENFA = ARGUMENTS.FA
        self.BACKWARDS = ARGUMENTS.BACKWARDS
        if ARGUMENTS.D == 'Firefox': 
            path_binary = r'./webdrivers/Firefox/linux_geckodriver' if platform == 'linux' else r"webdrivers\Firefox\geckodriver.exe" if platform == "win32" else r"webdrivers/Firefox/mac_geckodriver"
            options = FirefoxOptions()
            if ARGUMENTS.H: options.headless = True
            webdriver.Firefox.__init__(self, executable_path=path_binary, options=options)
        elif ARGUMENTS.D == 'Chrome': 
            path_binary = r'./webdrivers/Chrome/linux_chromedriver' if platform == 'linux' else r"webdrivers\Chrome\chromedriver.exe" if platform == "win32" else r"webdrivers/Chrome/mac_chromedriver"
            options = ChromeOptions()
            if ARGUMENTS.H: options.add_argument('headless')
            webdriver.Chrome.__init__(self, executable_path=path_binary, chrome_options=options)
        elif ARGUMENTS.D == 'Opera': 
            path_binary = r'./webdrivers/Opera/linux_operadriver' if platform == 'linux' else r"webdrivers\Opera\operadriver.exe" if platform == "win32" else r"webdrivers/Opera/mac_operadriver"
            options = OperaOptions()
            if ARGUMENTS.H: options.add_argument('headless')
            webdriver.Opera.__init__(self, executable_path=path_binary, options=options)
        elif ARGUMENTS.D == 'Edge': 
            if platform == "win32": 
                options = EdgeOptions()
                options.binary_location = r"webdrivers\Edge\msedgedriver.exe"
                if ARGUMENTS.H: options.headless = True
                webdriver.Edge.__init__(self, options)
            else: raise Exception("Edge is only supported on Windows ! ")
        else: 
            print('Unsupported driver, please use only "Firefox", "Opera", "Edge" or "Chrome".')
            raise Exception('BROWSER NOT SUPPORTED !')

        self.DiscordLogin(self.CHANNEL)
        self.SEND_MSG_BAR = self.FindOneByXPATH(self.SEND_MSG_BAR_XPATH, 'Discord sending bar')
        self.LAST_MSG = self.FindMessages('find last message')[0]

    def DiscordLogin(self, CHANNEL: str):
        self.get(CHANNEL)
        self.find_element_by_name("email").send_keys(self.MAIL)
        pwd = self.find_element_by_name("password")
        pwd.send_keys(self.PASSWD)
        pwd.send_keys(Keys.RETURN)
        if self.TOKENFA: input( f"Please type this code : '{ TOTP(self.TOKENFA).now() }' and then come back to this console to type enter." )

    def FindOneByCLASS(self, _class: str, item: str, 
                          timeout: int = 30):
        try:
            return WebDriverWait(self, timeout=timeout).until(lambda driver: driver.find_element_by_class_name(_class))
        except TimeoutException: raise Exception(f'"{ item }" took too much time to apear, dev can\'t really do anything to fix it')

    def FindOneByID(self, _id: str, item: str, 
                          timeout: int = 30):
        try:
            return WebDriverWait(self, timeout=timeout).until(EC.element_to_be_clickable((By.ID,_id)))
        except TimeoutException: raise Exception(f'"{ item }" took too much time to apear, dev can\'t really do anything to fix it')

    def FindOneByXPATH(self, PATH, item: str, 
                          timeout: int = 30):
        try:
            return WebDriverWait(self, timeout=timeout).until(EC.element_to_be_clickable((By.XPATH, PATH))) 
        except TimeoutException: raise Exception(f'"{ item }" took too much time to apear, dev can\'t really do anything to fix it')

    def FindMessages(self, item: str, 
                           timeout: int = 30):
        try:
            return WebDriverWait(self, timeout=timeout).until(lambda driver: driver.find_elements_by_xpath(self.MSG_XPATH))[::-1] 
        except TimeoutException: raise Exception(f'"{ item }" took too much time to apear, dev can\'t really do anything to fix it')

    def SendMessage(self, keys: str):
        if self.BACKWARDS : keys = keys[::-1]
        try: 
            self.SEND_MSG_BAR.send_keys(keys)
        except:
            self.SEND_MSG_BAR = self.FindOneByXPATH(self.SEND_MSG_BAR_XPATH, 'Discord sending bar')
            self.SEND_MSG_BAR.send_keys(keys)

        self.SEND_MSG_BAR.send_keys(Keys.RETURN)

    def CheckLast(self,
                  contains: str = ''):
        msgList = self.FindMessages('find all messages')
        for message in msgList[:msgList.index(self.LAST_MSG)]:
            try:
                sender, text = message.text.split('\n', 1)
                if self.POKEMEOW_USERNAME in sender and \
                   self.USERNAME in text and \
                   (all(contain in text for contain in contains) if type(contains) == list else contains in text):
                    self.LAST_MSG = message
                    return message 
            except: pass

    def WaitNew(self, to_send: str, contains, # contains can be a str or a list of str
                      retry: int = NBR_MAX_RECURSION, message_orig_txt = None, number_of_sleep: Decimal = Decimal(4) ):

        if not message_orig_txt: message_orig_txt = self.LAST_MSG.text
        # send le message
        self.SendMessage(to_send)

        # recupere une nouvelle fois le text du dernier embed
        new_message = self.CheckLast(contains)
        
        # tant que les deux messages sont egaux et que retry n'est pas passé
        while not new_message and number_of_sleep:
            sleep(0.1)
            number_of_sleep -= Decimal('0.1')
            new_message = self.CheckLast(contains)

        # si les deux embeds sont effectivement diffs
        if new_message and message_orig_txt != new_message.text: return new_message

        if retry :
            # sinon recur
            self.attente(6, 'trying again this action')
            return self.WaitNew(to_send, contains, retry - 1, message_orig_txt)
        else: raise RecursionError('MAX RECURSION REACHED')

    def attente(self, time: int, arg: str):
        for remaining in range(time, 0, -1):
            print(f"{remaining:2d}/{ time } seconds remaining before { arg }.", end = '\r')
            sleep(1)
        if platform != 'win32':
            stdout.write("\033[K")
            stdout.flush()
        else: print(' ' * 25)

    def WaitChangesOnMessage(self, to_send: str, message, 
                                   retry: int = NBR_MAX_RECURSION, number_of_sleep: Decimal = Decimal(4)) -> bool:
        bak_text = message.text
        self.SendMessage(to_send)
 
        while message.text == bak_text and number_of_sleep:
            sleep(0.1)
            number_of_sleep -= Decimal('0.1')

        if message.text != bak_text: return '\nCongratulations' in message.text

        if retry : 
            # sinon recur
            self.attente(6, 'trying again this action')
            return self.WaitChangesOnMessage(to_send, message, retry -1)
        else: raise RecursionError('MAX RECURSION REACHED')

    def Vote(self):
        print('Going to vote for you.')
        self.get(self.URL_LOGIN_TOP_GG)
        self.FindOneByXPATH(self.AUTHORIZE_BUTTON_XPATH, 'Discord Authorize button on top.gg').click()
        self.FindOneByCLASS('logo', 'Top.gg main logo')
        self.get(self.URL_VOTE_POKEMEOW)
        try: sleep(5); self.FindOneByXPATH("//a[@class='btn' and text() = 'No thanks']", 'No tanks button', 15).click()
        except: pass
        sleep(5)
        self.FindOneByXPATH(self.VOTE_BUTTON_XPATH, 'Vote button', 120).click()
        self.FindOneByID('rememberReminder', 'Vote has been validated', 120)
        self.get(self.CHANNEL)
        self.LAST_MSG = self.FindMessages('find last message')[0]
        print('I voted for you.')
